### ISSUE TITLE
Allow @ConfigurationProperties binding for immutable POJOs

### ISSUE THREAD

**Issue by sdeleuze** [spring-projects/spring-boot#8762]
Currently, it seems we are forced to use Kotlin classes with mutable nullable properties and default constructor with `@ConfigurationProperties` while idiomatic Kotlin code would be using classes with immutable properties initialized via constructor. I think there is a way to supporting that by leveraging `kotlin-reflect` library like [`jackson-module-kotlin`](https://github.com/FasterXML/jackson-module-kotlin) do.

More concretely, in MiXiT app I would like to be able to convert this [`MixitProperties`](https://github.com/mixitconf/mixit/blob/master/src/main/kotlin/mixit/MixitProperties.kt) class implementation to:

```
@ConfigurationProperties("mixit")
class MixitProperties(
    val baseUri: String,
    val admin: Credential,
    val drive: Drive
)

class Credential(
    val username: String,
    val password: String
)

class Drive(
    val fr: DriveDocuments,
    val en: DriveDocuments
)

class DriveDocuments(
    val sponsorform: String,
    val sponsor: String,
    val speaker: String,
    val press: String
)

```

We could imagine to support optional properties by using nullable types, like `val sponsorform: String?` for example.

I will be happy to help. I have also already worked with @apatrida who maintains Jackson Kotlin module, he may provide us some guidance I think.

**Comment by snicoll** [comment]
We currently use Spring Framework's binder (that's also used in the MVC world) and it does not support binding via constructor. I guess that's the problem you want to solve?

We're rewriting the binder for Boot 2.0, perhaps @philwebb has some plans about that?

**Comment by snicoll** [comment]
@jhoeller just pinged me to mention [SPR-15199](https://jira.spring.io/browse/SPR-15199). Would be pretty cool if we could use that somehow but I guess the best course of action is our binder and 2.0 anyway...

**Comment by apatrida** [comment]
There are two models that work with Kotlin.  The first, is that a plugin model to the binder let's Kotlin describe the constructors and properties providing information about which can be used to construct the class (or a combination of constructor + properties set after).  The second, is that a source of all settable things is made available and the decision is made by a Kotlin instantiator that does the work based on what it has available.  

The best support considers all of:

- what properties are set in the constructor
- what properties are writeable after construction
- for any property if it is nullable
- for any property in the constructor has a default value that can be used if it is absent

Kotlin reflection has a `callBy` method on the constructor that can be used in the case that default values are being used, otherwise it should use `call` method.

So the binder should give the opportunity to a plugin to do the introspection and instantiation so that the plugin can do the best job for the language. 

Binding examples for Kotlin:

Jackson Kotlin Module (pulls parameter name information, then hands it back along with actual parameters to a custom instantiator)

JDBI 3 (the Kotlin module handles all binding tasks when the target is a Kotlin class

**Comment by apatrida** [comment]
If there is a branch to track for the new binder, I can look at it and see what needs to be done to support Kotlin, and add that support.

**Comment by philwebb** [comment]
The experimental binder code is [here](https://github.com/philwebb/spring-boot/tree/env_conversion_hacking/), but it's still very much a work in progress. The plan is to make a binder that's specifically designed for binding from our configuration properties. It'll allow us to do some nice things like track the line/column number that a property was loaded from.

The [Binder](https://github.com/philwebb/spring-boot/blob/env_conversion_hacking/spring-boot/src/main/java/org/springframework/boot/context/properties/bind/Binder.java) class is pretty complete, and there are quite a few tests. It deals with Collections, Arrays, Maps and Scalar types directly. When it finds something it can't handle it delegates to [BeanBinder](https://github.com/philwebb/spring-boot/blob/env_conversion_hacking/spring-boot/src/main/java/org/springframework/boot/context/properties/bind/BeanBinder.java) implementations.

We currently have a `JavaBeanBinder` that works with mutable classes but the plan is to add a couple more. We'd like to have a `ValueObjectBeanBinder` that binds using constructor arguments (this should work well with Kotlin). 

We've also got plans for an `InterfaceBeanBinder`. The idea would be that you can so something like this:

```java
@ConfigurationProperties(prefix="mine")
public interface MyProperties {
 
    @DefaultValue("spring")
    public String getName();

    public int getAge();

}
```

Which would be used to generate a proxy class with the appropriate implementation.

The first step is to get the binder working with our existing mutable configuration property beans. We'll then look at extending the support.

**Comment by bclozel** [comment]
Quick update here: #8868 has been merged, the new binder is in master.

**Comment by sdeleuze** [comment]
@apatrida Since the new binder is now in master, any chance you could have a look to identify what needs to be done to provide first class support for immatable data classes in Kotlin?

**Comment by serandel** [comment]
Is this planned for the 2.0 release or for later in the roadmap? @ConfigurationProperties are a bit confusing without it.

**Comment by wilkinsona** [comment]
@serandel Can you please expand a bit on what you find confusing at the moment?

**Comment by serandel** [comment]
@wilkinsona Of course. :)

Compare this class from my project:

    @Component
    @ConfigurationProperties
    class ElasticsearchConfiguration {
        lateinit var origin : ElasticsearchSource
        var destination : ElasticsearchSource? = null
    
        override fun toString() = """
        |ElasticsearchConfiguration:
        |origin=$origin
        |destination=$destination
        """.trimMargin()
    }

to the version I tried first:

    @Component
    @ConfigurationProperties
    data class ElasticsearchConfiguration(val origin: ElasticsearchSource, val destination :ElasticsearchSource?)

Several features are dropped:

- properties have to be mutable
- properties have to be nullable or use lateinit
- equals, hashCode and toString are not included, and have to be written by hand (as I did in one case, just for debugging)
- copy and componentN are not included neither

From a Kotlin point of view, this type of configuration is obviously an immutable data class. There is some (very small, I concede) cognitive dissonance not being able to use a value type instead of a regular class for a limitation in the framework. Plus, the component is open to same rogue code altering a property.

I would really love to have this issue sorted out, as it would lead to beautiful, terse, expressive and idiomatic Kotlin data classes. :)

**Comment by serandel** [comment]
I just updated the previous comment. I thought that `lateinit var` would work... and it doesn't. :(

So, furthermore, we lose the distinction between optional and required properties. Perhaps it could be somewhat softened by using @required annotations and validating, I don't know, but it's not ideal.

**Comment by sdeleuze** [comment]
Indeed, from a Kotlin POV this should be an immutable class.

A regular Java `ValueObjectBeanBinder` that binds using constructor arguments should support Kotlin immutable classes (including data classes) out of the box. The main point to check will be how it behaves with null-safety.

A more advanced use case that will require specific Kotlin support will be classes with optional parameters and default values like this one since from a bytecode POV 2 constructors are generated:

```kotlin
@ConfigurationProperties
data class ElasticsearchConfiguration(
    val origin: ElasticsearchSource,
    val destination :ElasticsearchSource?,
    val foo: String = "bar"
)
```
I should be able to provide the relevant utils method as part of https://jira.spring.io/browse/SPR-15673 to deal with this use case, making possible for Spring Boot to leverage it.

**Comment by sdeleuze** [comment]
FYI Spring Framework 5.0 RC3 `BeanUtils` now fully supports Kotlin classes with immutable properties and optional parameters + default values while still providing API and implementation relevant for regular Java beans via the new [`BeanUtils#findPrimaryConstructor(Class)`](https://docs.spring.io/spring/docs/5.0.x/javadoc-api/org/springframework/beans/BeanUtils.html#findPrimaryConstructor-java.lang.Class-) method and [`BeanUtils.instantiateClass(Constructor, Object...)`](https://docs.spring.io/spring/docs/5.0.x/javadoc-api/org/springframework/beans/BeanUtils.html#instantiateClass-java.lang.reflect.Constructor-java.lang.Object...-) which now supports Kotlin classes with optional parameters + default values.

Notice that Spring Data [is about to add proper support for Kotlin constructor as well](https://github.com/spring-projects/spring-data-commons/pull/233), so such support in Spring Boot for `@ConfigurationProperties` is the last missing bit to get proper support for typical Spring + Kotlin application based on immutable POJOs which are idiomatic in Kotlin (and a best practice worth to promote IMO).

For what I have understood from previous discussion, and since we now have the confirmation that Kotlin support for immutable POJOs for `@ConfigurationProperties` is mainly a small extension of the Java support for that, should we modify the title and scope of this issue to "Allow `@ConfigurationProperties` binding for immutable POJOs" in order to have an equivalent of #1254 but applied to POJO binding?

I know you have a huge todo for Spring Boot 2.0, but with the guidance of somebody aware of the new `@ConfigurationProperties` infrastructure, maybe I could help in contributing Java and Kotlin immutable POJO PR? I guess that could also be a way to make sure that current design is a good fit with immutable classes and eventually adapt it before 2.0 API freeze.

**Comment by philwebb** [comment]
I think the code changes for the binding won't be that tricky. The annotation processor that generates the meta-data might be a bit harder.

**Comment by snicoll** [comment]
There are more side effect to its. Like annotations we can only put on the getter that we'll have to accept on fields. Non scalar (nested) properties is also a challenge as soon as it's not an inner class. Metadata generation will not work if `-parameters` isn't used.

**Comment by snicoll** [comment]
We discussed this one at our team call last week and it's still going to be a challenge in its current form. `@ConfigurationProperties` annotated types becomes beans ultimately. Either because the type was provided explicitly or because a bean is exposed manually.

Regardless, a bean has a lifecycle and the constructor is one of them. While we don't advocate for injecting anything in a `@ConfigurationProperties` object, there's nothing stopping you from doing that. Concretely, it means that we can't afford to use the constructor as a mean to pass the environment properties. 

That's why I see interface-based binding as a very good way forward: we keep control over the object's lifecyle and things are more consistent with what we have now. 

Of course, the other options would be to stop allowing a `@ConfigurationProperties` type to be a bean which is quite harsh and brings other challenges (binding of 3rd party object we build in an auto-config being the most obvious one).

**Comment by sdeleuze** [comment]
After discussing with @snicoll, I agree this feature requires to stop allowing a `@ConfigurationProperties` to be regular beans as well.

And in fact I see at least 3 advantages to remove such capability:
 - It would enable `@ConfigurationProperties` binding for immutable POJOs (solve this issue) which is nice to have in Java but the most idiomatic way to support such feature in Kotlin
 - It would make class and interface based `@ConfigurationProperties` bindings more consistent (in a sense interface based `@ConfigurationProperties` binding is a kind of workaround to avoid having the bean injection issue)
 - It would stop to confuse users by providing a more error-prone and opinionated solution (I see so many people confused by these double properties + bean injection capabilities on Stack Overflow).

I may be biased and there would be side effects as explained by @snicoll, but I think that would be a good change even if a breaking one, so I vote :+1: for removing such support from Boot 2.0.

To move forward about the original need for Kotlin + Boot developers regardless of what Boot team decide on this issue, my point of view is that the most important point is to provide a solution to the current situation where both escaped `@Value("\${foo}")` and nullable mutable setter/field based `@ConfigurationProperties` are 2 far from ideal possible solutions .

This can be solved by this issue (ideal solution) but also by [interface binding](https://github.com/spring-projects/spring-boot/issues/1254) if Kotlin interfaces with `val` properties are supported (I will had a comment on this issue with more details about Kotlin bytecode generated in such case in order to evaluate if such support will be doable or not).

**Comment by spencergibb** [comment]
Spring Cloud allows users to override `@Bean @ConfigurationProperties`. In other words, some are marked with `@ConditionalOnMissingBean`. We do have rare cases where we inject things into beans and take advantage of the spring bean lifecycle.

**Comment by wilkinsona** [comment]
How arduous would it be to update Cloud so that wasn't the case? We've decided in Boot that it was a mistake to allow people to override a `@ConfigurationProperties` beans. Overriding `ServerProperties` was particularly problematic, IIRC. It isn't explicitly prevented at the moment, but we're moving in that direction.

**Comment by spencergibb** [comment]
Unsure, the cases users have asked for are usually in places like providing hostnames and ipaddresses to the eureka instance configuration. We can certainly go in a different direction for Finchley if needed We inject config beans all over the place. What would be the mechanism if it wasn't a spring bean? @dsyer thoughts?

**Comment by wilkinsona** [comment]
`@EnableConfigurationProperties` will create a bean for all the classes listed. That's the recommended approach for creating a bean from a `@ConfigurationProperties` class

**Comment by dsyer** [comment]
As long as they have a bean lifecycle I don't think we should care how the bean definitions were created. It will mean some changes in Spring Cloud, but they ought to be cosmetic for users (as long as they don't override the `@Bean`).

**Comment by dsyer** [comment]
Just a reminder: we do still want to remind anything mutable (and I think that ought to be a Spring Boot feature but that's a separate topic). So we need to be able to identify all the beans that are `@ConfigurationProperties`.

**Comment by wlsc** [comment]
Any news on the immutable feature for ConfigurationProperties?

**Comment by philwebb** [comment]
@wlsc We'd still like to do it, but we're focusing on other 2.0 issues at the moment.

**Comment by wlsc** [comment]
@philwebb yes, @snicoll told me on twitter that is likely to be moved to 2.1 release.

**Comment by sdeleuze** [comment]
I have just seen that #1254 has been postponed to 2.1, I am sure for good reasons, but while providing a way to handle immutable and non-nullable configuration properties is probably nice to have in Java, this is from my point of view a critical issue in to solve in Boot 2.0 to provide a decent development experience with Kotlin. And that not just my own advice, a lot of developers working with Spring Boot + Kotlin gives that feedback.

After more thoughts and a discussion with @bclozel about that, I tend to think that trying to provide exactly the same feature for Java + Kotlin is a dead end. #1254 was not an ideal solution in Kotlin, that's maybe not a bad thing it has been postponed after all.

Previously, @snicoll accurately said:
> That's why I see interface-based binding as a very good way forward: we keep control over the object's lifecyle and things are more consistent with what we have now.

I do think [Kotlin data classes](https://kotlinlang.org/docs/reference/data-classes.htm) which are specifically designed to hold data, are suitable for that `@ConfigurationProperties`, even if in Java world interfaces are maybe the right answer. Also such restriction can make it easier to implement, since data classes are pure data and Spring Framework 5 provides now full support to instantiate such class.

My proposal, in order to provide a decent development experience in Kotlin with Boot 2.0, is to *provide support of `@ConfigurationProperties` limited to Kotlin data classes**. Kotlin `KClass` provides a `isData` property that could allow us to identify them easily and limit the support only to this kind of class.

I would like to contribute it via a pull-request with your guidance.

**Comment by sdeleuze** [comment]
Good news, while working on this issue, I found that the `noArg` plugin allows to use Kotlin mutable `@ConfigurationProperties` data classes with non-nullable properties.

Concretely, with the following configuration:

```kotlin
plugins {
  id "org.jetbrains.kotlin.plugin.noarg" version kotlinVersion
}
noArg {
    annotation("org.springframework.boot.context.properties.ConfigurationPro‌​perties")
}
```
It is possible to use the following mutable data classes:

```kotlin
@ConfigurationProperties("mixit")
data class MixitProperties(
    var baseUri: String,
    var contact: String,
    var drive: Drive,
    var aes: Aes) {

    data class Drive(
        var fr: DriveDocuments,
        var en: DriveDocuments)


    data class DriveDocuments(
        var sponsorform: String,
        var sponsor: String,
        var speaker: String,
        var press: String)

    data class Aes(
        var initvector: String,
        var key: String)
}
```
This issue is still needed to avoid the `noArg` trick and support immutable `@ConfigurationProperties`, but at least it provides a reasonably usable way to use `@ConfigurationProperties` with Kotlin + Spring Boot 2.0 regardless when we fix this issue in time for 2.0 or not.

In parallel, I am working on a Kotlin `BeanBinder` that supports immutable classes (including support for data classes) using Kotlin reflection. I will share the code later when I have something usable.

**Comment by sdeleuze** [comment]
Apology, the proposed no-arg workaround proposed above does not work with nest classes, it does not raise any error but nested classes are `null`, so I am back on working on Kotlin `BeanBinder` in high priority mode in order to get it work before Boot 2 RC1.

**Comment by asarkar** [comment]
@sdeleuze Boot 2 RC1 comes out tomorrow, is this feature going to be included?

**Comment by philwebb** [comment]
@asarkar No. This issue has no target milestone. It's unlikely we'll be able to get to it until 2.1.

**Comment by sdeleuze** [comment]
@serandel My latest tests with `lateinit var` seems to show it works as expected, could you please check again against Spring Boot 2.0.0.RC1 and Kotlin 1.2.20?

**Comment by ryan-barker-zefr** [comment]
Is there any current solution to this I've tried a bunch of different variations of the above and get various errors

**Comment by philwebb** [comment]
@ryan-barker-zefr No, we've not started working on this yet. We want to get it in Boot 2.1 if possible.

**Comment by sdeleuze** [comment]
@ryan-barker-zefr For now just use `lateinit var` as described in the [reference documentation](https://docs.spring.io/spring-boot/docs/current-SNAPSHOT/reference/htmlsingle/#boot-features-kotlin-configuration-properties).

**Comment by vojtapol** [comment]
Just a note of warning though, the metadata does not seem to be generated so there will be no autocomplete and no documentation unless I am missing something.

**Comment by snicoll** [comment]
@vojtapol thanks but that warning has nothing to do with the task at hand and we'd like to keep things focused. AFAIK, [kapt support](https://kotlinlang.org/docs/reference/kapt.html) works. If it does not, please create a separate issue with a sample that reproduces the issue.

**Comment by sdeleuze** [comment]
@vojtapol Even if that's surprising, [Kapt is not yet integrated in IDEA](https://youtrack.jetbrains.com/issue/KT-15040) so you need to build your project on command line (for example with `./gradlew clean build` if you use Gradle) to have your metadata generated.

**Comment by vojtapol** [comment]
Thank you @sdeleuze . Yes, I tried to manually invoke kapt through Maven in my case. It did generate the metadata file however only top-level properties were there. All nested properties were missing. Maybe it's a bug in kapt or in the kapt-maven-plugin.

**Comment by sdeleuze** [comment]
I have not tested with Maven, but with Gradle nested properties are generated correctly [in MiXiT project](https://github.com/mixitconf/mixit/blob/master/src/main/kotlin/mixit/MixitProperties.kt).

**Comment by vojtapol** [comment]
Finally figured out the problem of `kapt` only generating metadata with top-level properties. I will share it here in case it helps someone.

### Preconditions:
1. Use Maven
2. Have a mixed Kotlin + Java project, (`@ConfigurationProperties` is in Kotlin)
3. Have `spring-boot-configuration-processor` on the classpath
 

### The problem
In the [recommended Maven build setup for Kotlin ](https://kotlinlang.org/docs/reference/using-maven.html) `maven-compiler-plugin` is set to run after `kotlin-maven-plugin`. When the `kapt` goal on `kotlin-maven-plugin` runs the metadata is **correctly** generated.
Afterwards, `maven-compiler-plugin` runs its `compile` goal which also has annotation processing and if `spring-boot-configuration-processor` is on the classpath it will automatically discover it, run it and **OVERRIDE** the metadata with only top level properties.

### Solution:
Either
- somehow prevent `maven-compiler-plugin` from processing annotations 

or

- remove `spring-boot-configuration-processor` from dependencies

**Comment by vojtapol** [comment]
Finally figured out the problem of `kapt` only generating metadata with top-level properties. I will share it here in hopes it helps someone.

### Preconditions:
1. Use Maven
2. Have a mixed Kotlin + Java project, (`@ConfigurationProperties` file is written in Kotlin)
3. Have `spring-boot-configuration-processor` on the classpath


### The problem
In the [recommended Maven build setup for Kotlin ](https://kotlinlang.org/docs/reference/using-maven.html) `maven-compiler-plugin` is set to run after `kotlin-maven-plugin`. When the `kapt` goal on `kotlin-maven-plugin` runs the metadata is **correctly** generated.
Afterwards, `maven-compiler-plugin` runs its `compile` goal which also has annotation processing and if `spring-boot-configuration-processor` is on the classpath it will automatically discover it, run it and **OVERRIDE** the metadata with only top level properties.

### Solution:
Either
- somehow prevent `maven-compiler-plugin` from processing annotations

or

- remove `spring-boot-configuration-processor` from dependencies

**Comment by daggerok** [comment]
hell no, seems like much easier is use java for any ConfigurationProperties components...

**Comment by bclozel** [comment]
I'm wondering if the [JDK12 `record` JEP](https://cr.openjdk.java.net/~briangoetz/amber/datum.html) could be a nice match here for Java.

**Comment by vojtapol** [comment]
I am not sure if it's even needed anymore. Without `kapt` this will not work and `kapt` keeps breaking on every new `kotlin` release in our Maven build. Nobody from JetBrains is really maintaining it as far as I can tell. I am so tired of `kapt` issues I am thinking about just going back to Java POJOs for configuration.

**Comment by sdeleuze** [comment]
@vojtapol Could you provide a repro project of the issue you see in order to allow me to bring that to Kotlin team ?

**Comment by sdeleuze** [comment]
Thanks a lot Boot team for fixing that!

**Comment by wilkinsona** [comment]
Re-opening so that we can update the reference docs. I’d also like to take another look at the occasional need for `@Autowired` although I am not sure there’s a better alternative.

**Comment by wilkinsona** [comment]
The documentation refers to `@ConfigurationPropertyDefaultValue` but the annotation was renamed to `@DefaultValue`.

**Comment by Azbesciak** [comment]
In the [doc](https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-kotlin.html#boot-features-kotlin-configuration-properties) there is info that this is still not supported...

**Comment by snicoll** [comment]
That link is for the wrong version of Spring Boot.

This was updated properly: https://docs.spring.io/spring-boot/docs/2.2.0.BUILD-SNAPSHOT/reference/html/spring-boot-features.html#boot-features-kotlin-configuration-properties

**Comment by Azbesciak** [comment]
Ok, but this is first in the Google... and says that this is *current*.

**Comment by wilkinsona** [comment]
The `current` refers to the current, i.e. latest, Spring Boot release that is GA. Once 2.2 reaches that state, the `current` link will point to the documentation for 2.2.0.RELEASE.

**Comment by thekalinga** [comment]
@wilkinsona Does this work with Java aswell?

I tried with the following configuration & it did not resolve properties. Not sure if I had any mistake

`application.yml`

```yml
config:
  username: alpha
  password: password
  timeoutInSecs: 2
```

`ExternalSystemConfig.java`

```java
@Getter
@Builder
@AllArgsConstructor(onConstructor = @__(@JsonCreator))
@ToString
@ConfigurationProperties("config")
public class ExternalSystemConfig {
  private String username;
  private String password;
  private int timeoutInSecs;
}
```

`DemoApplication.java`

```java
@Log4j2
@SpringBootApplication
@EnableConfigurationProperties(ExternalSystemConfig.class)
public class DemoApplication {

  public static void main(String[] args) {
    new SpringApplicationBuilder()
        .sources(DemoApplication.class)
        .web(WebApplicationType.NONE)
        .run(args);
  }

  @Bean
  InitializingBean configLogger(ExternalSystemConfig config) {
    return () -> log.info("Config: {}",  config);
  }

}
```

But the `config` object that gets injected have all values set to null

I'm using `2.2.0.BUILD-SNAPSHOT` version for this

Any idea if I'm doing something wrong

Thanks!

**Comment by thekalinga** [comment]
If I make my object mutable, the binding works fine tho

**Comment by thekalinga** [comment]
This is how my generated class looks like

```java
  @ConstructorProperties({"username", "password", "timeoutInSecs"})
  @JsonCreator
  public ExternalSystemConfig(String username, String password, int timeoutInSecs) {
    this.username = username;
    this.password = password;
    this.timeoutInSecs = timeoutInSecs;
  }
```

**Comment by wilkinsona** [comment]
@thekalinga Immutable configuration property binding works with both Java and Kotlin. If you believe you've found a situation where it's not working as expected, please open a new issue with a complete and minimal sample that reproduces the problem. A project zipped and attached to the new issue or hosted in a separate GitHub repository is ideal.

**Comment by thekalinga** [comment]
@wilkinsona Just opened an issue with test project

https://github.com/spring-projects/spring-boot/issues/16928
